src/collector.rs

```diff
@@ -0,0 +1,128 @@
+//! The garbage collector.
+//!
+//! # Registration
+//!
+//! In order to track all handles in one place, we need some form of handle registration. When a
+//! handle is created, it is registered to a global lock-free singly-linked list of registries; and
+//! when a handle is dropped, it is unregistered from the list.
+
+use std::cmp;
+use std::ops::Deref;
+use std::sync::Arc;
+use std::sync::atomic::Ordering;
+use handle::{Handle, LocalEpoch, Scope, unprotected};
+use garbage::Bag;
+use epoch::Epoch;
+use sync::list::{List, Node};
+use sync::queue::Queue;
+
+
+/// A garbage collector.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_epoch as epoch;
+///
+/// let collector = epoch::Collector::new();
+///
+/// let handle = collector.add_handle();
+/// handle.pin(|scope| {
+///     scope.flush();
+/// });
+/// ```
+#[derive(Debug)]
+pub struct Collector(Arc<Global>);
+
+/// The global data for a garbage collector.
+#[derive(Debug)]
+pub struct Global {
+    /// The head pointer of the list of handle registries.
+    registries: List<LocalEpoch>,
+    /// A reference to the global queue of garbages.
+    garbages: Queue<(usize, Bag)>,
+    /// A reference to the global epoch.
+    epoch: Epoch,
+}
+
+impl Collector {
+    pub fn new() -> Self {
+        Self { 0: Arc::new(Global::new()) }
+    }
+
+    pub fn add_handle(&self) -> Handle {
+        Handle::new(&self.0)
+    }
+}
+
+impl Deref for Collector {
```

Comment:
Do we ever use this impl? If not, it should probably be removed because it's showing up in the list of implemented traits in the documentation, but isn't useful to a user because `Global` isn't exposed, so they couldn't do anything with it.

Reply:
That's the idea, yeah. But I don't think we need expose `push_bag` in `Collector`. `push_bag` is just an implementation detail, not a publicly visible function.

`collect` would be fine to expose, though, but we should give more thought to the interface. Here's a quick suggestion what a public `collect` function might look like:

```rust
fn collect() -> Collect {
    // ...
}

enum Collect {
    Empty, // No more garbage to collect.
    HasMore, // There is more garbage. You may call `collect()` again if you want.
    BlockedByPin, // There is more garbage, but it can't be collected at the moment.
}
```

Reply:
I pushed a commit on the new comments.  In retrospect, it would have been better to put this PR as an RFC first.  If we still has unresolved issues, I'd like to make an RFC based on this branch..

- I made `Collector::collect()` unsafe, because you need to ensure that no concurrent handle is accessing the objects in the garbage queue.

- For the return type of `collect()`, I'd like to make it a separate PR or RFC, as this PR is growing larger and larger.